chain/chain/src/chain.rs

```diff
@@ -1767,26 +1771,79 @@ impl Chain {
         shard_id: ShardId,
         sync_hash: CryptoHash,
         num_parts: u64,
-    ) -> Result<(), Error> {
+        state_parts_task_scheduler: &Box<dyn Fn(StatePartsMessage)>,
+        prev_execution_result: &Option<StatePartsResponse>,
+        source: StatePartsTaskSource,
+    ) -> Result<SetStateFinalizeResult, Error> {
         let shard_state_header = self.get_state_header(shard_id, sync_hash)?;
         let mut height = shard_state_header.chunk_height_included();
         let state_root = shard_state_header.chunk_prev_state_root();
         let epoch_id = self.get_block_header(&sync_hash)?.epoch_id().clone();
+        let map_key = (epoch_id.clone(), shard_id);
+        let current_task = self.currently_applying_state_part.get(&map_key);
 
-        for part_id in 0..num_parts {
-            let key = StatePartKey(sync_hash, shard_id, part_id).try_to_vec()?;
-            let part = self.store.owned_store().get(ColStateParts, &key)?.unwrap();
+        if let Some(result) = prev_execution_result {
+            if result.shard_id != shard_id || result.epoch_id != epoch_id {
+                // Results are not for this shard
+                return Ok(SetStateFinalizeResult::NoAction);
+            }
+            match current_task {
+                Some(scheduled_part_id) => {
+                    if *scheduled_part_id != result.part_id {
+                        // We had probably had to abort process last time
+                        return Ok(SetStateFinalizeResult::NoAction);
+                    }
+                }
+                None => {
+                    // We had probably had to abort process last time
+                    return Ok(SetStateFinalizeResult::NoAction);
+                }
+            }
```

Comment:
These cases seem operationally interesting, makes sense to `tracing::info` them.